### DISCUSSION TITLE
Applying a formula to an image (Curves... again)

### DISCUSSION THREAD

**Discussion by Build-0-Matic** [ImageMagick/ImageMagick#7317] [Digital Image Processing] [answered]
Hi, I feel that this probably a very redundant question, but it has me stumped.
I want to apply an adjustment curve to an image. So I calculated my curve profile so that it best matches what the curve would look like in photoshop. I arrived at this y = - 0.00345 * x^{2} + 0.63314 * x + 2.26039
So, since I have a lot of images to work on at the same time, I tried the following Imagemagick command line
`magick mogrify -format jpg -fx "-0.00345*u^(2) + 0.63314*u + 2.26039" "%temp%\*.jpg"`
But all I get is an empty file...

I know I'm doing something wrong as I don't fully understand how this thing works (I'm not a programmer by trade)

Anybody has any idea how to fix this?

Here's the Imagemagick -version info
Version: ImageMagick 7.1.1-21 Q16-HDRI x64 58d86e4:20231021 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Channel-masks(64-bit) Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2022 (193532217)

**Comment by fmw42** [comment]
Imagemagick -fx expressions assume u,v etc are image coordinates in the range 0 to 1.  Are you coefficient scaled appropriately for that range?  If you are on a Unix-like system or can use Window 10/11 unix or Cygwin, then you could try my script, curves, at http://www.fmwconcepts.com/imagemagick/curves/index.php to do a spline fit to a list of control points.  The points can be in the range 0 to 100 (percent) or 0 to 255 for 8-bit values or whatever you want as controlled by the scale arguments.

**Reply by Build-0-Matic** [reply]
I don't know if that changes much, but since the measured points I had were a bit all over the place, I had to go here (https://tools.softinery.com/CurveFitter/) to get the formula calculated (I haven't done this by hand in so long that I kinda forgot). So the values should be scaled (if not, I just need to divide by 100)

**Reply by fmw42** [reply]
Then use the HALD approach below.

**Comment by fmw42** [comment]
If you want to apply the same mapping to all images, you could process it once using HALD images and then apply it to many image using -hald-clut.  See https://imagemagick.org/Usage/color_mods/#hald-clut.  You can create the original HALD image and transfer it to Photoshop. The process the HALD image with the mapping you like. Then save that result and bring it back to Imagemagick and use it with -hald-clut to process many images.

**Reply by Build-0-Matic** [reply]
So there is no easy way around to fix that... bummer.

**Comment by fmw42** [comment]
If you are using coordinates for your points in the 8-bit range from Photoshop, then divide the values by 255 and do your fit again to get the 2nd order coefficients that you want for Imagemagick -fx equation.

**Reply by Build-0-Matic** [reply]
I'm currently thinking of using Libreoffice calc to reproduce a set of coordinate values, taking the answers and putting them through your curves script and using that answer... would that work?

**Reply by fmw42** [reply]
Post your list of control points and an image you want to process. Let me see how varied they are and try my script or scale them and also do the scaling and fit them to second order for your -fx process. I can't work on it until later this afternoon or evening.  You can just proceed and try yourself while I am out, if you want.

Have you tried the hald process? It is very easy.

**Reply by Build-0-Matic** [reply]
I Haven't tried Hald yet... I'm currently multitasking between three jobs... not fun

**Reply by Build-0-Matic** [reply]
Here is the list of points measured points. Input,Output
1,0 1,5 7,10 34,15 37,20 68,25 70,30 85,40 91,50 96,60 96,70 97,80 98,90 98,100

This is for making the adjustment curve for an old 1860's photographic process (cyanotype) which has a very shallow possible density range. Just throw at it any B&W photo and it should compress the tonal ranges.

The thing with the Hald process is that it needs the companion file to go with it. I just don't want to have too many things to pack with the batch script I'm working on.

It's funny because going into this project I wasn't expecting it to be so complicated 🙄

**Comment by snibgo** [comment]
> -fx "-0.00345\*u^(2) + 0.63314\*u + 2.26039"

Values of `u` will be in the range of 0 to 1. When u==0, the output value will be exactly 2.26039, on a nominal range of 0 to 1. I don't suppose this is what you intend.

Perhaps the formula you give is suitable for values in the range 0 to 255. If so, then try this:
```
-fx "ubig=u*255; (-0.00345*ubig^(2) + 0.63314*ubig + 2.26039)/255"
```

**Comment by snibgo** [comment]
> Here is the list of points measured points. Input,Output
1,0 1,5 7,10 34,15 37,20 68,25 70,30 85,40 91,50 96,60 96,70 97,80 98,90 98,100

No simple formula will map one input value (such as 1) to multiple output values (such as 0 and 5).

**Reply by Build-0-Matic** [reply]
I just tried refining the fitment some more using another (better) site and found that a cubic formula fits best the physical material I'm using.
the suggested formula is
y=(4.034e-04*x^3)+(-4.908e-02*x^2)+(1.777e+00*x)+1.913e-01

That is for values from 0 to 100

**Reply by fmw42** [reply, marked as answer]
So using your formula with snibgo's method on the lena gray image, I get

![lena_gray](https://github.com/ImageMagick/ImageMagick/assets/13458218/3cfbb4a3-bac3-466d-b3fe-67917d442c30)

`magick lena_gray.png -fx "uu=u*100; (4.034e-04*uu*uu*uu -4.908e-02*uu*uu + 1.777*uu + 1.913e-01)/100" lena_gray_cubic.png
`
![lena_gray_cubic](https://github.com/ImageMagick/ImageMagick/assets/13458218/8654f4a0-5167-41fa-9f9e-9fb8d8fc872d)

You could also use uu^3 in place of uu*uu*uu. Similarly for uu^2 in place of uu*uu.  I am not sure which would be faster.

**Reply by Build-0-Matic** [reply]
As for speed, I feel it would be just a marginal difference. So I probably wouldn't bother optimizing this any more.
So, if I recap to see if I correctly understand what is going on: 
u is the pixel value
uu=u*100 is there allow more digits in the values.
Then comes the formula which gets it's result divided by 100 to counteract the first *100.
Am I getting this one right?

**Reply by fmw42** [reply]
Not quite, but mostly correct.

uu=u*100 is because Imagemagick has u values between 0 and 1, but you are providing values between 0 and 100. So your formula was for values between 0 and 100. So we must multiply u by 100 to account for your values and your coefficients. Later we have to divide by 100 to get the results back to a range 0 to 1 for the output.  Imagemagick then scales it back automatically to the raw values for your version Q level (8-bit or 16-bit values).

**Reply by Build-0-Matic** [reply]
Ah! OK, now I get it.
It feels a bit strange to work with values between 0 and 1. But hey, if it works, that's all we ask of it.
Really many thanks to you. Looking at this, I'm sure I would still be stumbling around to get this working 😊

**Comment by fmw42** [comment]
I loaded your points. Assuming they mean percent in range 0 to 100 as per the values you provided, then when using the curve fitting tool at https://planetcalc.com/8735/

[pts.txt](https://github.com/ImageMagick/ImageMagick/files/15327760/pts.txt)

![Screenshot 2024-05-15 at 3 26 30 PM](https://github.com/ImageMagick/ImageMagick/assets/13458218/b9dc3a10-3da3-4edd-b6e0-ca28de22b727)

![Screenshot 2024-05-15 at 3 27 36 PM](https://github.com/ImageMagick/ImageMagick/assets/13458218/eabc6ec1-b241-432e-9da8-34c81ee40767)

Then using -fx on a grayscale version of the lena image using snibgo's approach

![lena_gray](https://github.com/ImageMagick/ImageMagick/assets/13458218/d2731ba2-5d15-4a80-a816-930253ad2245)


`magick lena_gray.png -fx "uu=u/100; 0.0119*uu*uu - 0.4725*uu + 8.5427" lena_gray_fx.png`

gives me a totally white output.

**Comment by fmw42** [comment]
OK, this time I used my approach to divide the control points by 100.

[cpts.txt](https://github.com/ImageMagick/ImageMagick/files/15327735/cpts.txt)


![Screenshot 2024-05-15 at 4 15 55 PM](https://github.com/ImageMagick/ImageMagick/assets/13458218/d89c74f9-20a0-4c1f-b4a9-9f189b675047)

and got this fit

![Screenshot 2024-05-15 at 4 16 23 PM](https://github.com/ImageMagick/ImageMagick/assets/13458218/8ec92c62-9afc-4099-ab39-0c2910beb748)

with this 2nd order equation

y=1.1271x2−0.3851x+0.0614

`magick lena_gray.png -fx "uu=1.127*u*u - 0.3851*u + 0.0614" lena_gray_fx3.png`

The result on the lena image was

![lena_gray_fx3](https://github.com/ImageMagick/ImageMagick/assets/13458218/1eab2152-89c8-4631-9689-7eea0ec575ac)

**Comment by snibgo** [comment]
> Then using -fx on a grayscale version of the lena image using snibgo's approach
magick lena_gray.png -fx "uu=u/100; 0.0119\*uu\*uu - 0.4725\*uu + 8.5427" lena_gray_fx.png
gives me a totally white output.

If the formula is suited for a range 0 to 100, then we need to multiply (not divide) u by 100, then divide the result by 100.
```
magick lena_gray.png -fx "uu=u*100; (0.0119*uu*uu - 0.4725*uu + 8.5427)/100" lena_gray_fx.png
```
This gives the same result as your:
```
y=1.1271x2−0.3851x+0.0614
```

**Reply by fmw42** [reply]
OOPS on my part!

**Reply by fmw42** [reply]
I do not see how you convert your fx expression by doing what you say and get my same coefficients as mine. Multiply and dividing by 100 should only change the coefficient decimal place, not the values of the digits. Can you clarify?  Or are you just saying if you process the image with your fx expression, you get the same resulting image as mine.

**Comment by snibgo** [comment]
>  Or are you just saying if you process the image with your fx expression, you get the same resulting image as mine.

Yes. These two commands give virtually the same result:

```
magick toes.png -fx "uu=u*100; (0.0119*uu*uu - 0.4725*uu + 8.5427)/100" f.png

magick toes.png -fx "1.1271*uu*uu - 0.3851*uu + 0.0614" f2.png
```
We can simplify the first of these, to avoid dividing by 100 at every pixel:
```
magick toes.png -fx "uu=u*100; (0.000119*uu*uu - 0.004725*uu + 0.085427)" f3.png
```
